### PR TITLE
fix css scoping

### LINKED ISSUE #2763: Inconsistent behavior with Javascript when running app from an iframe
### Describe the bug

If someone wants to select things on a Gradio app using Javascript, the behavior changes if they are in an iframe or outside of an iframe. Outside of an iframe, a `shadowRoot` is required to select the element. But when inside an iframe, having a `shadowRoot` doesn't select the element and breaks the app.

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

[An example](https://huggingface.co/spaces/multimodalart/shadow-root/blob/main/app.py#L19) using `shadowRoot`: 
- Works when it is direct access/localhost, like here: https://multimodalart-shadow-root.hf.space/
- Breaks when it is an iframe/Space, like here:
https://huggingface.co/spaces/multimodalart/shadow-root

[An example](https://huggingface.co/spaces/multimodalart/shadow-root-reverse/blob/main/app.py#L19) *not* using `shadowRoot`
- Works when it is an iframe/Space, like here: https://huggingface.co/spaces/multimodalart/shadow-root-reverse
- Breaks when it is direct access/localhost, like here: https://multimodalart-shadow-root-reverse.hf.space/

This makes the AUTOMATIC1111 WebUI incompatible with running on Hugging Face Spaces (or any other iframe) for e.g., as they manipulate the app with Javascript with a shadowRoot: https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/master/script.js#L2 

### System Info

```shell
Gradio == 3.12.0
```

### Severity

Except if there's a universal way to select elements we are missing, it is blocking all usage of gradio inside a Space if Javascript is manipulated

### LINKED ISSUE #3036: Option to disable shadow dom
- [ ✔] I have searched to see if a similar issue already exists.

**Is your feature request related to a problem? Please describe.**  
Feels like many developers are just struggling to add custom features to gradio components because the shadow dom prevent javascript from reaching the components easily. I've searched a lot in the issues but none of them allows accessing shadow dom. It took me a whole day trying to figure out how to add a click event to the gallery with external javascript (added to the closing of body) but ended up deciding to switch to another framework...

**Describe the solution you'd like**  
Just simply add an option to disable the shadow dom and make all html elements visible to external javascript will be much helpful. 

**Additional context**  
If it's not possible to disable shadow dom, I would also appreciate that if we can have an interface to load js into the whole app (instead of for individual components) or allow directly editing the html of the coponents.

### PR DESCRIPTION
**Pull request by pngwn**
# Description

Ignore the branch name. This PR fixes the weird unstyled Gradio we have been seeing on the Docs.

The actual reason for this is kinda weird but essentially it only happened when there was more than one instance of gradio on the page, embedded via the web component.

What this PR does is removes the use of the shadow DOM and mounts the CSS + JS into the head of the parent document, sharing the styles across all instances of `gradio-app` rather than each `gradio-app` needing their own set of styles. We can do this now because svelte is handling most of the scoping for us. We _do_ still need to scope css resets and the few global styles we have. I have done this by prefixing all non hashed (svelte) css selectors with `gradio-container-<VERSION>` (except`.` has been converted to `-`), so something like `gradio-container-3-16-1 h1 { ... }`. This means you could potentially have multiple different versions of gradio on the same page and they wouldn't conflict.

This also simplifies the code somewhat, we needed multiple build time modifications in order to make things work with custom elements and the shadow DOM. Since they only worked during build and not dev we needed to mount the app differently in different modes. These distinctions are now gone.

tldr: we're scoping differently.

closes #3036. probably closes #2763.

Please include: 
* relevant motivation
* a summary of the change 
* which issue is fixed. 
* any additional dependencies that are required for this change.

Closes: # (issue)

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have added a short summary of my change to the CHANGELOG.md
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

### CODE REVIEW

**Review by aliabd — APPROVED** [review]
Tested on [staging](https://www.gradio-staging.com/docs/main) and everything works as expected! 

Small note on min-height css. You'll see the embedded component demos have css that overrides min-height and and removes the footers so they don't take unnecessary space. Here's an [example](https://huggingface.co/spaces/gradio/textbox_component/blob/main/run.py#L3). Should I change that selector to `gradio-container-3-16-2`? That's what i'm seeing now. 

Also all that type of css will need to be much more specific now without the shadow DOM. For example all footers on every demo are gone. The docs page footer is gone too! Not a big deal though, I can clean this up after the PR is merged.

**Comment by abidlabs** [comment]
Does this mean that if a user updates their version of gradio, they would have to change any custom `css` that they have passed into the Blocks/Interface? That would be a pretty big inconvenience imo

**Comment by pngwn** [comment]
Good point. I'll change this so that there is both the version specific class name (for our own scoping) in addition to a normal `gradio-container` (which is easier for users).

**Comment by pngwn** [comment]
This PR would mean that there is an increased chance of collisions, it is technically possible to keep the shadow DOM and fix this issue but quite difficult due to shortcomings in the web component API. The API that facilitates this is actually supported in chrome but not safari. 

In order to keep things fully scoped (with the shadow DOM) but fix the issues we are facing it would take a bit of time because some of our tooling would need to be rewritten. I can take that approach if we are uncomfortable with the risk this change would introduce.

**Comment by abidlabs** [comment]
I actually think removing the shadow DOM is a good idea for our users as well (it's caused lots of confusion / questions in the past). My main concern was users having to change the CSS between versions -- if that's not going to the case, this sounds great!

**Comment by pngwn** [comment]
@aliabd Can you test this again now. I've added back the `gradio-container` class so that should styling a bit easier, although as the shadow DOM is gone anyone embedding will need to be careful not to  write styles that impact the embedded gradio app. Gradio itself will never affect the containing page though.

**Comment by aliabd** [comment]
Just tested on staging @pngwn everything looks good: https://www.gradio-staging.com/docs/main

**Comment by abidlabs** [comment]
Nice fix @pngwn! One nit: can we update the documentation on embedding with web components to include this caveat when using your own `css`. 

Relevant file: https://github.com/gradio-app/gradio/blob/main/guides/01_getting-started/03_sharing-your-app.md#embedding-with-web-components

**Comment by pngwn** [comment]
@abidlabs I have added a note to that guide, explaining the caveat and which kinds of selectors are more likely to cause issues. If there is a better way to phrase or a better place to put it then let me know!

**Comment by abidlabs** [comment]
LGTM @pngwn thanks for the awesome PR!

**Review by abidlabs — APPROVED** [review]

**Comment by SoulflareRC** [comment]
Is this fix added in 3.17? I'm very excited with the removal of shadow DOM! With 3.16.2 I'm still having trouble running the reload mode as #3057 describes. I can only develop with 3.15...

**Comment by abidlabs** [comment]
Hi @SoulflareRC yes it's added in 3.17 which was just released. Let us know if you are still having any issues.

**Comment by SoulflareRC** [comment]
Is it possible to add custome js and css after this update? I was trying to do so by modifying the template response but neither jquery or js seem to be able to access any gradio component.

**Comment by pngwn** [comment]
Nothing should changed regarding custom CSS or JS. Could you open a new issue or discussion or post on discord with some code?
